File: src/core/orchestrator.py
"""
Orchestrator - Coordena fluxo completo CREATE e IMPLEMENT.
"""
import asyncio
from pathlib import Path
from typing import Optional, Literal
from dataclasses import dataclass
from rich.console import Console
from src.core.model_router import ModelRouter
from src.agents.planner import PlannerAgent
from src.agents.coder import CoderAgent
from src.agents.tester import TesterAgent
from src.services.scaffolder import ScaffolderService
from src.adapters.repo_manager import RepoManager

console = Console()

@dataclass
class CreateResult:
    path: Path
    success: bool
    message: str

@dataclass
class ImplementResult:
    success: bool
    pr_url: Optional[str]
    message: str

class MainOrchestrator:
    """Orquestrador principal do AI Dev Agent."""
    
    def __init__(self, config):
        self.config = config
        self.router = ModelRouter(config)
        self.planner = PlannerAgent(self.router)
        self.coder = CoderAgent(self.router)
        self.tester = TesterAgent(self.router)
        self.scaffolder = ScaffolderService(self.router, config)
        self.repo_manager = RepoManager(config)
    
    async def create_project(
        self,
        project_type: str,
        name: str,
        language: str,
        framework: Optional[str] = None,
        database: Optional[str] = None,
        include_auth: bool = False
    ) -> CreateResult:
        """Cria novo projeto do zero."""
        try:
            with console.status("[cyan]Criando projeto...[/cyan]"):
                # 1. Gerar estrutura
                project_path = await self.scaffolder.generate_structure(
                    project_type=project_type,
                    name=name,
                    language=language,
                    framework=framework,
                    database=database,
                    include_auth=include_auth
                )
                
                # 2. Inicializar Git
                self.repo_manager.init_repo(project_path)
                
                # 3. Commit inicial
                self.repo_manager.commit(project_path, "chore: initial commit")
                
                console.print(f"[green]✅ Projeto criado: {project_path}[/green]")
                
                return CreateResult(
                    path=project_path,
                    success=True,
                    message=f"Projeto {name} criado com sucesso!"
                )
        
        except Exception as e:
            return CreateResult(
                path=Path(),
                success=False,
                message=f"Erro ao criar projeto: {str(e)}"
            )
    
    async def implement_feature(
        self,
        issue_url: Optional[str] = None,
        text: Optional[str] = None,
        repo_url: Optional[str] = None
    ) -> ImplementResult:
        """Implementa feature em projeto existente."""
        try:
            # 1. Clone repo
            with console.status("[cyan]Clonando repositório...[/cyan]"):
                repo_path = self.repo_manager.clone_repo(repo_url)
            
            # 2. Criar plano
            with console.status("[cyan]Criando plano...[/cyan]"):
                demand = text if text else f"Issue: {issue_url}"
                plan = await self.planner.create_plan(demand, str(repo_path))
                
                console.print("\n[bold cyan]📋 Plano de Implementação:[/bold cyan]")
                console.print(f"  Título: {plan.get('title', 'N/A')}")
                console.print(f"  Subtarefas: {len(plan.get('subtasks', []))}")
            
            # 3. Implementar
            with console.status("[cyan]Implementando código...[/cyan]"):
                changes = await self.coder.implement_plan(plan, repo_path)
                console.print(f"[green]✅ {len(changes)} arquivos modificados[/green]")
            
            # 4. Gerar testes
            with console.status("[cyan]Gerando testes...[/cyan]"):
                test_files = await self.tester.generate_tests(changes, repo_path)
                console.print(f"[green]✅ {len(test_files)} arquivos de teste criados[/green]")
            
            # 5. Commit e PR
            branch_name = f"feature/{plan.get('title', 'implementation').lower().replace(' ', '-')}"
            self.repo_manager.create_branch(repo_path, branch_name)
            self.repo_manager.commit(repo_path, f"feat: {plan.get('title', 'implementation')}")
            
            pr_url = self.repo_manager.create_pull_request(
                repo_url=repo_url,
                title=plan.get('title', 'Feature Implementation'),
                body=plan.get('description', ''),
                branch=branch_name
            )
            
            return ImplementResult(
                success=True,
                pr_url=pr_url,
                message="Feature implementada com sucesso!"
            )
        
        except Exception as e:
            return ImplementResult(
                success=False,
                pr_url=None,
                message=f"Erro: {str(e)}"
            )

File: src/core/orchestrator.ts
import * as path from 'path';
import chalk from 'chalk';
import ora from 'ora';
import { AppConfig, CreateResult, ImplementResult } from '../types';
import { ModelRouter } from './modelRouter';
import { PlannerAgent } from '../agents/planner';
import { CoderAgent } from '../agents/coder';
import { TesterAgent } from '../agents/tester';
import { ScaffolderService } from '../services/scaffolder';
import { RepoManager } from '../adapters/repoManager';
import { GitHubProvider } from '../adapters/githubProvider';

export class MainOrchestrator {
  private router: ModelRouter;
  private planner: PlannerAgent;
  private coder: CoderAgent;
  private tester: TesterAgent;
  private scaffolder: ScaffolderService;
  private repoManager: RepoManager;
  private github: GitHubProvider;

  constructor(private config: AppConfig) {
    this.router = new ModelRouter(config);
    this.planner = new PlannerAgent(this.router);
    this.coder = new CoderAgent(this.router);
    this.tester = new TesterAgent(this.router);
    this.scaffolder = new ScaffolderService(this.router, config);
    this.repoManager = new RepoManager(config);
    this.github = new GitHubProvider(config.githubToken);
  }

  async createProject(
    projectType: string,
    name: string,
    language: string,
    framework?: string,
    database?: string,
    includeAuth: boolean = false,
  ): Promise<CreateResult> {
    const spinner = ora('Creating project...').start();
    try {
      const projectPath = await this.scaffolder.generateStructure({
        projectType,
        name,
        language,
        framework,
        database,
        includeAuth,
      });

      await this.repoManager.initRepo(projectPath);
      await this.repoManager.commit(projectPath, 'chore: initial commit');

      spinner.succeed(chalk.green(`Project created: ${projectPath}`));
      return { path: projectPath, success: true, message: `Project ${name} created successfully!` };
    } catch (err) {
      spinner.fail(chalk.red('Failed to create project'));
      return { path: '', success: false, message: `Error: ${(err as Error).message}` };
    }
  }

  async implementFeature(
    repoUrl: string,
    issueUrl?: string,
    text?: string,
  ): Promise<ImplementResult> {
    let issueNumber: number | undefined;

    try {
      // 1. Resolve demand from issue or text
      let demand: string;
      if (issueUrl) {
        const spinner = ora('Fetching GitHub issue...').start();
        try {
          const issue = await this.github.getIssue(issueUrl);
          demand = `${issue.title}\n\n${issue.body}`;
          issueNumber = issue.number;
          spinner.succeed(`Issue #${issue.number}: ${issue.title}`);
        } catch (err) {
          spinner.fail('Failed to fetch issue');
          throw err;
        }
      } else if (text) {
        demand = text;
      } else {
        return { success: false, message: 'Provide either --issue or --text' };
      }

      // 2. Clone
      let spinner = ora('Cloning repository...').start();
      let repoPath: string;
      try {
        repoPath = await this.repoManager.cloneRepo(repoUrl);
        spinner.succeed('Repository cloned');
      } catch (err) {
        spinner.fail('Failed to clone repository');
        throw err;
      }

      // 3. Plan
      spinner = ora('Creating implementation plan...').start();
      let plan;
      try {
        plan = await this.planner.createPlan(demand, repoPath);
        spinner.succeed(`Plan: ${plan.title} (${plan.subtasks.length} subtasks)`);
      } catch (err) {
        spinner.fail('Failed to create plan');
        throw err;
      }

      console.log(chalk.cyan('\nPlan:'));
      plan.subtasks.forEach((t, i) => console.log(`  ${i + 1}. ${t.title}`));

      // 4. Implement
      spinner = ora('Implementing code...').start();
      let changes;
      try {
        changes = await this.coder.implementPlan(plan, repoPath);
        spinner.succeed(chalk.green(`${changes.length} files modified`));
      } catch (err) {
        spinner.fail('Failed to implement code');
        throw err;
      }

      // 5. Tests
      spinner = ora('Generating tests...').start();
      try {
        const testFiles = await this.tester.generateTests(changes, repoPath);
        spinner.succeed(chalk.green(`${testFiles.length} test files created`));
      } catch (err) {
        spinner.warn('Test generation failed, continuing...');
      }

      // 6. Commit + Push + PR
      spinner = ora('Creating branch and PR...').start();
      try {
        const branchName = `feature/${plan.title.toLowerCase().replace(/\s+/g, '-').substring(0, 50)}`;
        await this.repoManager.createBranch(repoPath, branchName);
        await this.repoManager.commit(repoPath, `feat: ${plan.title}`);

        // Push branch to remote so GitHub API can find it
        await this.repoManager.pushBranch(repoPath, repoUrl, branchName);

        const { owner, repo } = GitHubProvider.parseRepoUrl(repoUrl);
        const prUrl = await this.github.createPullRequest({
          owner,
          repo,
          title: plan.title,
          body: `## Summary\n${plan.description}\n\n## Changes\n${plan.subtasks.map((t) => `- ${t.title}`).join('\n')}`,
          head: branchName,
          base: this.config.defaultBaseBranch,
          draft: this.config.createDraftPr,
          issueNumber,
        });

        spinner.succeed(chalk.green('PR created'));
        return { success: true, prUrl, message: 'Feature implemented successfully!' };
      } catch (err) {
        spinner.fail('Failed to create PR');
        throw err;
      }
    } catch (err) {
      return { success: false, message: `Error: ${(err as Error).message}` };
    }
  }
}
